### PR TITLE
[JSC] PropertyTable should have compact mode

### PR DESCRIPTION
**Pull request by Constellation**
#### 29dc23b5224f5b9343d79365649b59f22525a513
<pre>
[JSC] PropertyTable should have compact mode
<a href="https://bugs.webkit.org/show_bug.cgi?id=239451">https://bugs.webkit.org/show_bug.cgi?id=239451</a>

Reviewed by Saam Barati.

This patch introduces &quot;compact&quot; version of PropertyTable. PropertyTable needs to hold two kind of integers: index and property offset.
But for most of objects, both are pretty small and they can fit in uint8_t. If we can use uint8_t for both, we can significantly reduce
size of allocated memory for PropertyTable (only 40% of memory is required!). This is good for memory, but also good for performance.
Now each CompactPropertyTableEntry is 8bytes while original PropertyMapEntry was 16bytes, so CompactPropertyTableEntry can fit in CPU cache well.
Also, not allocating large amount of memory can reduce memory allocation / deallocation cost. One of costly destruction of GC-managed objects is
PropertyTable (and CodeBlock), and we can reduce that cost by not allocating much memory.

The approach is following.

1. For index vector, we use uint8_t if index can fit within uint8_t.
2. For proprety offset, we use uint8_t and CompactPropertyTableEntry if it is suitable.
3. Once the table gets non-compact, we keep it non-compact since we could have deleted index which has larger than uint8_t. We could improve this
   strategy when deleted indexes are cleared, but for now, we are taking simple approach.
4. We store isCompactFlag 1 bit in the pointer to the table.
5. We encapsulate functions modifying property table entry in PropertyTable itself, so we do not leak internal compact / non-compact mode to the user
   of PropertyTable. We remove begin() / end() iterators and instead use forEachproperty, which can implement iteration for each mode more efficiently.

We have a further opportunity to improve this further: we can deploy 75% load factor only for compact table. Then we can increase threshold of
compact table further and keep more and more tables compact mode. Plus, for small sized tables, small backing memory is better in terms of
CPU cache hit (and that&apos;s measured in WTF::HashTable, and that&apos;s why WTF::Hashtable deploys 75% load factor only for small tables). This is left
for the subsequent change.

This change is neutral in JetStream2, 0.3% improvement in Speedometer2 with 80% confidence, and 0.41% improvement in RAMification with 95% confidence.

* JSTests/stress/change-attribute-structure-transition.js:
(shouldBe.JSON.stringify.sd):
* Source/JavaScriptCore/bytecode/ObjectAllocationProfileInlines.h:
(JSC::ObjectAllocationProfileBase&lt;Derived &gt;::initializeProfile):
* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter&lt;AbstractStateType &gt;::executeEffects):
* Source/JavaScriptCore/dfg/DFGConstantFoldingPhase.cpp:
(JSC::DFG::ConstantFoldingPhase::foldConstants):
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
* Source/JavaScriptCore/ftl/FTLOperations.cpp:
(JSC::FTL::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/runtime/ClonedArguments.h:
* Source/JavaScriptCore/runtime/IteratorOperations.cpp:
(JSC::createIteratorResultObjectStructure):
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):
* Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSONObject.cpp:
(JSC::Stringifier::Holder::appendNextProperty):
* Source/JavaScriptCore/runtime/JSObject.cpp:
(JSC::JSObject::analyzeHeap):
* Source/JavaScriptCore/runtime/JSObject.h:
* Source/JavaScriptCore/runtime/ObjectConstructor.h:
(JSC::constructEmptyObject):
(JSC::createDataPropertyDescriptorObjectStructure):
(JSC::createAccessorPropertyDescriptorObjectStructure):
* Source/JavaScriptCore/runtime/ObjectConstructorInlines.h:
(JSC::objectAssignFast):
* Source/JavaScriptCore/runtime/PropertyOffset.h:
* Source/JavaScriptCore/runtime/PropertySlot.h:
* Source/JavaScriptCore/runtime/PropertyTable.cpp:
(JSC::PropertyTable::PropertyTable):
(JSC::PropertyTable::finishCreation):
(JSC::PropertyTable::visitChildrenImpl):
(JSC::PropertyTable::~PropertyTable):
(JSC::PropertyTable::seal):
(JSC::PropertyTable::freeze):
(JSC::PropertyTable::isSealed const):
(JSC::PropertyTable::isFrozen const):
(JSC::PropertyTable::renumberPropertyOffsets):
* Source/JavaScriptCore/runtime/PropertyTable.h:
(JSC::isPowerOf2):
(JSC::nextPowerOf2):
(JSC::PropertyTable::findImpl):
(JSC::PropertyTable::find):
(JSC::PropertyTable::get):
(JSC::PropertyTable::add):
(JSC::PropertyTable::remove):
(JSC::PropertyTable::take):
(JSC::PropertyTable::updateAttributeIfExists):
(JSC::PropertyTable::sizeInMemory):
(JSC::PropertyTable::reinsert):
(JSC::PropertyTable::rehash):
(JSC::PropertyTable::skipDeletedEntries):
(JSC::PropertyTable::dataSize):
(JSC::PropertyTable::canInsert):
(JSC::PropertyTable::forEachProperty const):
(JSC::PropertyTable::forEachPropertyMutable):
(JSC::PropertyTable::begin): Deleted.
(JSC::PropertyTable::end): Deleted.
(JSC::PropertyTable::begin const): Deleted.
(JSC::PropertyTable::end const): Deleted.
(JSC::PropertyTable::table): Deleted.
(JSC::PropertyTable::table const): Deleted.
* Source/JavaScriptCore/runtime/RegExpMatchesArray.h:
* Source/JavaScriptCore/runtime/Structure.cpp:
(JSC::Structure::dumpStatistics):
(JSC::Structure::Structure):
(JSC::Structure::materializePropertyTable):
(JSC::Structure::nonPropertyTransitionSlow):
(JSC::Structure::isSealed):
(JSC::Structure::isFrozen):
(JSC::Structure::flattenDictionaryStructure):
(JSC::PropertyTableStatisticsExitLogger::~PropertyTableStatisticsExitLogger):
(JSC::Structure::getConcurrently):
(JSC::Structure::getPropertiesConcurrently):
(JSC::Structure::getPropertyNamesFromStructure):
(JSC::Structure::toStructureShape):
(JSC::Structure::dump const):
* Source/JavaScriptCore/runtime/Structure.h:
(JSC::CompactPropertyTableEntry::CompactPropertyTableEntry):
(JSC::CompactPropertyTableEntry::key const):
(JSC::CompactPropertyTableEntry::setKey):
(JSC::CompactPropertyTableEntry::offset const):
(JSC::CompactPropertyTableEntry::setOffset):
(JSC::CompactPropertyTableEntry::attributes const):
(JSC::CompactPropertyTableEntry::setAttributes):
(JSC::PropertyTableEntry::PropertyTableEntry):
(JSC::PropertyTableEntry::key const):
(JSC::PropertyTableEntry::setKey):
(JSC::PropertyTableEntry::offset const):
(JSC::PropertyTableEntry::setOffset):
(JSC::PropertyTableEntry::attributes const):
(JSC::PropertyTableEntry::setAttributes):
(JSC::PropertyMapEntry::PropertyMapEntry): Deleted.
* Source/JavaScriptCore/runtime/StructureInlines.h:
(JSC::Structure::get):
(JSC::Structure::forEachPropertyConcurrently):
(JSC::Structure::forEachProperty):
(JSC::Structure::add):
(JSC::Structure::remove):
(JSC::Structure::attributeChange):

Canonical link: <a href="https://commits.webkit.org/249881@main">https://commits.webkit.org/249881@main</a>
git-svn-id: <a href="https://svn.webkit.org/repository/webkit/trunk@293210">https://svn.webkit.org/repository/webkit/trunk@293210</a> 268f45cc-cd09-0410-ab3c-d52691b4dbfc
</pre>

### CODE REVIEW

File: Source/JavaScriptCore/ChangeLog [hunk 1]
```diff
@@ -1,3 +1,140 @@
...
+        Reviewed by NOBODY (OOPS!).
+
+        This patch introduces "compact" version of PropertyTable. PropertyTable needs to hold two kind of integers: index and property offset.
+        But for most of objects, both are pretty small and they can fit in uint8_t. If we can use uint8_t for both, we can significantly reduce
+        size of allocated memory for PropertyTable (only 40% of memory is required!). This is good for memory, but also good for performance.
+        Now each CompactPropertyTableEntry is 8bytes while original PropertyMapEntry was 16bytes, so CompactPropertyTableEntry can fit in CPU cache well.
+        Also, not allocating large amount of memory can reduce memory allocation / deallocation cost. One of costly destruction of GC-managed objects is
+        PropertyTable (and CodeBlock), and we can reduce that cost by not allocating much memory.
+
+        The approch is the followings.
```
Comment (saambarati) [inline Source/JavaScriptCore/ChangeLog]: nit: followings => following

File: Source/JavaScriptCore/runtime/PropertyTable.cpp [hunk 2]
```diff
@@ -59,28 +59,32 @@ PropertyTable::PropertyTable(VM& vm, unsigned initialCapacity)
...
     , m_indexSize(sizeForCapacity(initialCapacity))
     , m_indexMask(m_indexSize - 1)
-    , m_index(static_cast<unsigned*>(PropertyTableMalloc::zeroedMalloc(dataSize())))
+    , m_indexVector()
     , m_keyCount(0)
     , m_deletedCount(0)
 {
     ASSERT(isPowerOf2(m_indexSize));
+    bool isCompact = tableCapacity() < UINT8_MAX;
+    m_indexVector = (bitwise_cast<uintptr_t>(PropertyTableMalloc::zeroedMalloc(dataSize(isCompact))) | (isCompact ? isCompactFlag : 0));
```
Comment (saambarati) [inline Source/JavaScriptCore/runtime/PropertyTable.cpp]: nit: No parens needed around bitwise_cast

File: Source/JavaScriptCore/runtime/PropertyTable.cpp [hunk 3]
```diff
@@ -59,28 +59,32 @@ PropertyTable::PropertyTable(VM& vm, unsigned initialCapacity)
...
+    ASSERT(isCompact == this->isCompact());
 }
 
 PropertyTable::PropertyTable(VM& vm, const PropertyTable& other)
     : JSCell(vm, vm.propertyTableStructure.get())
     , m_indexSize(other.m_indexSize)
     , m_indexMask(other.m_indexMask)
-    , m_index(static_cast<unsigned*>(PropertyTableMalloc::malloc(dataSize())))
+    , m_indexVector(bitwise_cast<uintptr_t>(PropertyTableMalloc::malloc(dataSize(other.isCompact()))) | (other.isCompact() ? isCompactFlag : 0))
```
Comment (saambarati) [inline Source/JavaScriptCore/runtime/PropertyTable.cpp]: We make m_indexVector in two places. Maybe we can add a helper that takes size and isCompact as arguments, and does the tagging? That way the logic is all in one place.

File: Source/JavaScriptCore/runtime/PropertyTable.cpp [hunk 4]
```diff
@@ -59,28 +59,32 @@ PropertyTable::PropertyTable(VM& vm, unsigned initialCapacity)
...
     , m_keyCount(other.m_keyCount)
     , m_deletedCount(other.m_deletedCount)
 {
     ASSERT(isPowerOf2(m_indexSize));
+    ASSERT(isCompact() == other.isCompact());
+    memcpy(bitwise_cast<void*>(m_indexVector & indexVectorMask), bitwise_cast<void*>(other.m_indexVector & indexVectorMask), dataSize(isCompact()));
```
Comment (saambarati) [inline Source/JavaScriptCore/runtime/PropertyTable.cpp]: `m_indexVector & indexVectorMask` happens in a lot of places in this patch. Can we add a helper for that? Maybe `indexVector()`?

File: Source/JavaScriptCore/runtime/PropertyTable.cpp [hunk 5]
```diff
@@ -92,18 +96,34 @@ PropertyTable::PropertyTable(VM& vm, unsigned initialCapacity, const PropertyTab
...
     , m_indexMask(m_indexSize - 1)
-    , m_index(static_cast<unsigned*>(PropertyTableMalloc::zeroedMalloc(dataSize())))
+    , m_indexVector()
     , m_keyCount(0)
     , m_deletedCount(0)
 {
     ASSERT(isPowerOf2(m_indexSize));
     ASSERT(initialCapacity >= other.m_keyCount);
+    bool isCompact = other.isCompact() && tableCapacity() < UINT8_MAX;
+    m_indexVector = (bitwise_cast<uintptr_t>(PropertyTableMalloc::zeroedMalloc(dataSize(isCompact))) | (isCompact ? isCompactFlag : 0));
```
Comment (saambarati) [inline Source/JavaScriptCore/runtime/PropertyTable.cpp]: that helper I suggested can be used here too

File: Source/JavaScriptCore/runtime/PropertyTable.cpp [hunk 6]
```diff
@@ -136,12 +156,103 @@ void PropertyTable::destroy(JSCell* cell)
...
+    });
+    clearDeletedOffsets();
+    return offset;
+}
+
+template<typename Functor>
+inline void PropertyTable::forEachPropertyMutable(const Functor& functor)
+{
+    uintptr_t indexVector = m_indexVector;
+    if (indexVector & isCompactFlag) {
```
Comment (saambarati) [inline Source/JavaScriptCore/runtime/PropertyTable.cpp]: can this just use isCompact() and access the vector using the helper I suggested?

File: Source/JavaScriptCore/runtime/PropertyTable.cpp [hunk 7]
```diff
@@ -92,18 +96,34 @@ PropertyTable::PropertyTable(VM& vm, unsigned initialCapacity, const PropertyTab
...
+        });
+    } else {
+        auto* indexVector = bitwise_cast<uint32_t*>(m_indexVector & indexVectorMask);
+        auto* table = tableFromIndexVector(indexVector);
+        other.forEachProperty([&](auto& entry) {
+            ASSERT(canInsert(entry));
+            reinsert(indexVector, table, entry);
+            entry.key()->ref();
+            return IterationStatus::Continue;
+        });
```
Comment (saambarati) [inline Source/JavaScriptCore/runtime/PropertyTable.cpp]: Do we really need the branch here? The two sides are the same, minus how we iterate. But how we iterate is already handled by `forEachProperty`. So I don't think we need this branch with duplicated code. The lambda should just handle it since it's using `auto`

File: Source/JavaScriptCore/runtime/PropertyTable.h [hunk 8]
```diff
@@ -78,52 +78,34 @@ inline unsigned nextPowerOf2(unsigned v)
...
+// offset is small enough (64 currently), so that we can still assume that most of property offsets are < 256.
+//
+// 1. If property offset gets larger than 255, then we get non-compact PropertyTable. It requires at least 191 (255 - 64) properties.
+//    In that case, PropertyTable size should be 256 since it is power-of-two.
+// 2. If index gets larger than 255, then we get non-compact PropertyTable. But we are using 0 and 255 for markers. Thus, if we get 253
+//    used counts, then we need to change the table.
+//
+// So, typical scenario is that, once 128th property is added, then we extend the table via rehashing. At that time, we change the
+// table from compact to non-compact mode. We could explore the way to keep this more larger, like, up to 192 by using 75% load factor
+// for small table (including compact table).
```
Comment (saambarati) [inline Source/JavaScriptCore/runtime/PropertyTable.h]: This last sentence is a bit confusing. I know you're exploring this already, maybe we can just leave the comment out for now

File: Source/JavaScriptCore/runtime/PropertyTable.h [hunk 9]
```diff
@@ -238,182 +213,240 @@ class PropertyTable final : public JSCell {
...
+
+    bool isCompact() const { return m_indexVector & isCompactFlag; }
+
+    template<typename Functor>
+    void forEachPropertyMutable(const Functor&);
+
+    // The table of values lies after the hash index.
+    CompactPropertyTableEntry* tableFromIndexVector(uint8_t* index)
+    {
+        return reinterpret_cast_ptr<CompactPropertyTableEntry*>(index + m_indexSize);
```
Comment (saambarati) [inline Source/JavaScriptCore/runtime/PropertyTable.h]: nit: Let's use `bitwise_cast` for these functions

File: Source/JavaScriptCore/runtime/PropertyTable.h [hunk 10]
```diff
@@ -238,182 +213,240 @@ class PropertyTable final : public JSCell {
...
-
-inline PropertyTable::const_iterator PropertyTable::end() const
-{
-    auto* tableEnd = this->tableEnd();
-    return const_iterator(tableEnd, tableEnd);
-}
-
-inline PropertyTable::find_iterator PropertyTable::find(const KeyType& key)
+template<typename Index, typename Entry>
+std::tuple<unsigned, unsigned, PropertyOffset, uint8_t> PropertyTable::findImpl(const Index* indexVector, const Entry* table, const KeyType& key)
```
Comment (saambarati) [inline Source/JavaScriptCore/runtime/PropertyTable.h]: nit: Can we make a simple struct here instead of tuple. tuple with 4 elements is hard to follow compared to

```
struct FindResult {
    unsigned entryIndex;
    unsigned index;
    PropertyOffset offset;
    unsigned attributes;
};

File: Source/JavaScriptCore/runtime/PropertyTable.h [hunk 11]
```diff
@@ -238,182 +213,240 @@ class PropertyTable final : public JSCell {
...
     }
 }
 
-inline PropertyTable::ValueType* PropertyTable::get(const KeyType& key)
+inline std::tuple<unsigned, unsigned, PropertyOffset, uint8_t> PropertyTable::find(const KeyType& key)
+{
+    ASSERT(key);
+    ASSERT(key->isAtom() || key->isSymbol());
+    uintptr_t indexVector = m_indexVector;
+    if (indexVector & isCompactFlag) {
```
Comment (saambarati) [inline Source/JavaScriptCore/runtime/PropertyTable.h]: nit: isCompact()?

File: Source/JavaScriptCore/runtime/PropertyTable.h [hunk 12]
```diff
@@ -238,182 +213,240 @@ class PropertyTable final : public JSCell {
...
+        const auto* index = bitwise_cast<const uint8_t*>(indexVector & indexVectorMask);
```
Comment (saambarati) [inline Source/JavaScriptCore/runtime/PropertyTable.h]: Might even be worth adding a template parameter to `indexedVector` getter I suggested. By default, it's `void*`, but it can be passed `uint8_t*` and `uint32_t*`, etc

File: Source/JavaScriptCore/runtime/PropertyTable.h [hunk 13]
```diff
@@ -238,182 +213,240 @@ class PropertyTable final : public JSCell {
...
     }
 
     // Allocate a slot in the hashtable, and set the index to reference this.
-    unsigned entryIndex = usedCount() + 1;
-    m_index[iter.second] = entryIndex;
-    iter.first = &table()[entryIndex - 1];
-    *iter.first = entry;
+    entryIndex = usedCount() + 1;
+    uintptr_t indexVector = m_indexVector;
+    if (indexVector & isCompactFlag) {
```
Comment (saambarati) [inline Source/JavaScriptCore/runtime/PropertyTable.h]: `isCompact()`?

File: Source/JavaScriptCore/runtime/PropertyTable.h [hunk 14]
```diff
@@ -217,10 +190,12 @@ class PropertyTable final : public JSCell {
     void finishCreation(VM&);
 
     // Used to insert a value known not to be in the table, and where we know capacity to be available.
-    void reinsert(const ValueType& entry);
+    template<typename Index, typename Entry>
+    void reinsert(Index*, Entry*, const ValueType& entry);
 
     // Rehash the table. Used to grow, or to recover deleted slots.
-    void rehash(VM&, unsigned newCapacity);
+    static bool canBeFitInCompact(const ValueType& entry) { return entry.offset() <= UINT8_MAX; }
```
Comment (saambarati) [inline Source/JavaScriptCore/runtime/PropertyTable.h]: name nit: `canBeFitInCompact` => `canFitInCompact`

File: Source/JavaScriptCore/runtime/PropertyTable.h [hunk 15]
```diff
@@ -238,182 +213,240 @@ class PropertyTable final : public JSCell {
...
+        ASSERT_UNUSED(offset, offset == invalidOffset);
+        ASSERT_UNUSED(entryIndex, entryIndex == EmptyEntryIndex);
     }
 
     // Allocate a slot in the hashtable, and set the index to reference this.
-    unsigned entryIndex = usedCount() + 1;
-    m_index[iter.second] = entryIndex;
-    iter.first = &table()[entryIndex - 1];
-    *iter.first = entry;
+    entryIndex = usedCount() + 1;
```
Comment (saambarati) [inline Source/JavaScriptCore/runtime/PropertyTable.h]: in the isCompact case, can we assert `usedCount() < UINT8_MAX`? I also think entryIndex here is `unsigned`, so alternatively, assert that it's <= UINT8_MAX. Since this works by canInsert check above.

File: Source/JavaScriptCore/runtime/PropertyTable.h (hunk 13 above)
Comment (saambarati) [reply Source/JavaScriptCore/runtime/PropertyTable.h]: I guess alternatively, we could have a function like:

```
template <typename Function>
void withIndexVector(Function function) {
    if (isCompact()) function(indexVector<uint8_t*>();
    else function(indexVector<uint32_t*>();
}
```
And we don't need to repeat the same code in both sides of the branch

File: Source/JavaScriptCore/runtime/PropertyTable.h [hunk 16]
```diff
@@ -238,182 +213,240 @@ class PropertyTable final : public JSCell {
...
     ++propertyTableStats->numRemoves;
 #endif
 
     // Replace this one element with the deleted sentinel. Also clear out
     // the entry so we can iterate all the entries as needed.
-    m_index[iter.second] = deletedEntryIndex();
-    iter.first->key->deref();
-    iter.first->key = PROPERTY_MAP_DELETED_ENTRY_KEY;
+    uintptr_t indexVector = m_indexVector;
+    if (indexVector & isCompactFlag) {
```
Comment (saambarati) [inline Source/JavaScriptCore/runtime/PropertyTable.h]: If you like my `withIndexVector`, it can be used here too, since both sides of the branch are the same minus type of index vector

File: Source/JavaScriptCore/runtime/PropertyTable.h [hunk 17]
```diff
@@ -238,182 +213,240 @@ class PropertyTable final : public JSCell {
...
+    }
+    auto* vector = bitwise_cast<uint32_t*>(indexVector & indexVectorMask);
+    auto* table = tableFromIndexVector(vector);
+    auto [entryIndex, index, offset, oldAttributes] = findImpl(vector, table, key);
+    UNUSED_VARIABLE(index);
+    UNUSED_VARIABLE(oldAttributes);
+    if (offset == invalidOffset)
+        return invalidOffset;
+    table[entryIndex - 1].setAttributes(attributes);
+    return offset;
```
Comment (saambarati) [inline Source/JavaScriptCore/runtime/PropertyTable.h]: nit: This can also use my withIndexVector suggestion.

File: Source/JavaScriptCore/runtime/PropertyTable.h [hunk 18]
```diff
@@ -479,59 +512,104 @@ inline PropertyTable* PropertyTable::copy(VM& vm, unsigned newCapacity)
...
-
-    for (; iter != end; ++iter) {
-        ASSERT(canInsert());
-        reinsert(*iter);
+    // Once table gets non-compact, we do not change it back to compact again.
+    // This is because some of property offset can be larger than UINT8_MAX already.
+    bool isCompact = !forceNonCompact && oldIsCompact && tableCapacity() < UINT8_MAX;
+    size_t newDataSize = dataSize(isCompact);
+    uintptr_t indexVector = bitwise_cast<uintptr_t>(PropertyTableMalloc::zeroedMalloc(newDataSize));
+    m_indexVector = (indexVector | (isCompact ? isCompactFlag : 0));
```
Comment (saambarati) [inline Source/JavaScriptCore/runtime/PropertyTable.h]: nit: my makeIndexedVector suggestion can be used here too

File: Source/JavaScriptCore/runtime/PropertyTable.h [hunk 19]
```diff
@@ -479,59 +512,104 @@ inline PropertyTable* PropertyTable::copy(VM& vm, unsigned newCapacity)
...
-    m_index[iter.second] = entryIndex;
-    table()[entryIndex - 1] = entry;
+    indexVector[index] = entryIndex;
+    table[entryIndex - 1] = entry;
 
     ++m_keyCount;
 }
 
-inline void PropertyTable::rehash(VM& vm, unsigned newCapacity)
+inline void PropertyTable::rehash(VM& vm, unsigned newCapacity, bool forceNonCompact)
```
Comment (saambarati) [inline Source/JavaScriptCore/runtime/PropertyTable.h]: nit: we could invert this bool and call it something like `canStayCompact` so we don't have double negatives in the code below

File: Source/JavaScriptCore/runtime/PropertyTable.h [hunk 20]
```diff
@@ -479,59 +512,104 @@ inline PropertyTable* PropertyTable::copy(VM& vm, unsigned newCapacity)
...
-        ASSERT(canInsert());
-        reinsert(*iter);
+    // Once table gets non-compact, we do not change it back to compact again.
+    // This is because some of property offset can be larger than UINT8_MAX already.
+    bool isCompact = !forceNonCompact && oldIsCompact && tableCapacity() < UINT8_MAX;
+    size_t newDataSize = dataSize(isCompact);
+    uintptr_t indexVector = bitwise_cast<uintptr_t>(PropertyTableMalloc::zeroedMalloc(newDataSize));
+    m_indexVector = (indexVector | (isCompact ? isCompactFlag : 0));
+
+    if (isCompact) {
```
Comment (saambarati) [inline Source/JavaScriptCore/runtime/PropertyTable.h]: I wonder if these 3 different cases can be abstracted into a lambda that just takes auto* for old and new vector

File: Source/JavaScriptCore/runtime/PropertyTable.h [hunk 21]
```diff
@@ -579,9 +647,40 @@ inline unsigned PropertyTable::sizeForCapacity(unsigned capacity)
...
+        const auto* index = bitwise_cast<const uint32_t*>(indexVector & indexVectorMask);
+        const auto* cursor = tableFromIndexVector(index);
+        const auto* end = tableEndFromIndexVector(index);
+        for (; cursor != end; ++cursor) {
+            if (cursor->key() == PROPERTY_MAP_DELETED_ENTRY_KEY)
+                continue;
+            if (functor(*cursor) == IterationStatus::Done)
+                return;
+        }
+    }
```
Comment (saambarati) [inline Source/JavaScriptCore/runtime/PropertyTable.h]: nit: Seems like another good candidate for my `withIndexVector` function

**Review by saambarati — APPROVED** [review]
Nice. r=me. Your algorithm looks solid, and the patch is nice. I just have a few ideas on how to clean up the code. I think we can remove a lot of repeated code by adding some helpers.

File: Source/JavaScriptCore/ChangeLog (hunk 1 above)
Comment (Constellation) [reply Source/JavaScriptCore/ChangeLog]: Fixed.

File: Source/JavaScriptCore/runtime/PropertyTable.cpp (hunk 2 above)
Comment (Constellation) [reply Source/JavaScriptCore/runtime/PropertyTable.cpp]: Fixed.

File: Source/JavaScriptCore/runtime/PropertyTable.h (hunk 13 above)
Comment (Constellation) [reply Source/JavaScriptCore/runtime/PropertyTable.h]: Sounds good! Introduced this.

File: Source/JavaScriptCore/runtime/PropertyTable.h (hunk 21 above)
Comment (Constellation) [reply Source/JavaScriptCore/runtime/PropertyTable.h]: Done.

File: Source/JavaScriptCore/runtime/PropertyTable.h (hunk 20 above)
Comment (Constellation) [reply Source/JavaScriptCore/runtime/PropertyTable.h]: I've just used withIndexVector twice, and probably, because of how isCompact is induced, I think compiler can drop one of that case.

File: Source/JavaScriptCore/runtime/PropertyTable.h (hunk 16 above)
Comment (Constellation) [reply Source/JavaScriptCore/runtime/PropertyTable.h]: Nice, done.

File: Source/JavaScriptCore/runtime/PropertyTable.h (hunk 12 above)
Comment (Constellation) [reply Source/JavaScriptCore/runtime/PropertyTable.h]: I rewrote these things with `withIndexVector`.

File: Source/JavaScriptCore/runtime/PropertyTable.h (hunk 11 above)
Comment (Constellation) [reply Source/JavaScriptCore/runtime/PropertyTable.h]: Ditto. I changed it with `withIndexVector`.

File: Source/JavaScriptCore/runtime/PropertyTable.h (hunk 19 above)
Comment (Constellation) [reply Source/JavaScriptCore/runtime/PropertyTable.h]: Changed.

File: Source/JavaScriptCore/runtime/PropertyTable.h (hunk 18 above)
Comment (Constellation) [reply Source/JavaScriptCore/runtime/PropertyTable.h]: Done.

File: Source/JavaScriptCore/runtime/PropertyTable.h (hunk 17 above)
Comment (Constellation) [reply Source/JavaScriptCore/runtime/PropertyTable.h]: Done.

File: Source/JavaScriptCore/runtime/PropertyTable.h (hunk 15 above)
Comment (Constellation) [reply Source/JavaScriptCore/runtime/PropertyTable.h]: Done.

File: Source/JavaScriptCore/runtime/PropertyTable.h (hunk 14 above)
Comment (Constellation) [reply Source/JavaScriptCore/runtime/PropertyTable.h]: Done

File: Source/JavaScriptCore/runtime/PropertyTable.cpp (hunk 4 above)
Comment (Constellation) [reply Source/JavaScriptCore/runtime/PropertyTable.cpp]: I just changed it to `withIndexVector`.

File: Source/JavaScriptCore/runtime/PropertyTable.cpp (hunk 3 above)
Comment (Constellation) [reply Source/JavaScriptCore/runtime/PropertyTable.cpp]: Done.

File: Source/JavaScriptCore/runtime/PropertyTable.cpp (hunk 5 above)
Comment (Constellation) [reply Source/JavaScriptCore/runtime/PropertyTable.cpp]: Done.

File: Source/JavaScriptCore/runtime/PropertyTable.cpp (hunk 6 above)
Comment (Constellation) [reply Source/JavaScriptCore/runtime/PropertyTable.cpp]: Done.

File: Source/JavaScriptCore/runtime/PropertyTable.cpp (hunk 7 above)
Comment (Constellation) [reply Source/JavaScriptCore/runtime/PropertyTable.cpp]: Changed it to use `withIndexVector`.

File: Source/JavaScriptCore/runtime/PropertyTable.h (hunk 8 above)
Comment (Constellation) [reply Source/JavaScriptCore/runtime/PropertyTable.h]: OK, dropped.

File: Source/JavaScriptCore/runtime/PropertyTable.h (hunk 9 above)
Comment (Constellation) [reply Source/JavaScriptCore/runtime/PropertyTable.h]: Done.

File: Source/JavaScriptCore/runtime/PropertyTable.h (hunk 10 above)
Comment (Constellation) [reply Source/JavaScriptCore/runtime/PropertyTable.h]: Done

**Review by Constellation — COMMENTED** [review]
Thanks!

**Comment by Constellation** [comment]
Landed https://commits.webkit.org/249881@main (29dc23b5224f)!
